redhat-developer/service-binding-operator
# Default Projection as Files

### Motivation

- Fixes #893 
- By default, set `bindAsFiles` as `true`.

### Changes

- Makes `BindAsFiles` field a boolean pointer.

### Behaviour 

**[DEFAULT BEHAVIOUR]** If `bindAsFiles` is not set by the user, the operator sets `bindAsFiles` as `true` and the service credentials are injected as files.

If `bindAsFiles` is set as `true` by the user, the service credentials are injected as files.

If `bindAsFiles` is set as `false` by the user,the service credentials are injected as environment variables.

### Manual Testing for Default Behaviour

1. Follow the steps given in this [example](https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_postgresql) to:
- install DB operator, Service Binding Operator
- Import an application
- Create a DB instance 

2. Create the following `ServiceBindingRequest`:

```
      apiVersion: binding.operators.coreos.com/v1alpha1
      kind: ServiceBinding
      metadata:
          name: binding-request
      spec:
          application:
              name: nodejs-app
              group: apps
              version: v1
              resource: deployments
          services:
          -   group: postgresql.baiju.dev
              version: v1alpha1
              kind: Database
              name: db-demo
```

Notice `bindAsFiles` flag is not set.

3. After ServiceBinding's status condition `Ready` is set to `True`, ssh into the application pod by running command:
`k exec -it <pod_name> -- bash`

E.g. `k exec -it nodejs-app-67bdf56459-vfxqt -- bash`

The command prompt will change. 

4. Execute the following command to find the binding files

```
bash-5.0# cd bindings/binding-request
```

5. Ensure service environment variables are not present in the list of environment variables by executing the following command:
```
bash-5.0# printenv
```

### Acceptance Test

The above scenario is tested in acceptance tests. Execute `make test-acceptance` to run the acceptance tests.

The default behaviour i.e. when `bindAsFiles` is not set by the user is tested in this acceptance test -> https://github.com/redhat-developer/service-binding-operator/pull/907/files#diff-87d32cff8bb83e390f5f95e2d617292442f15ace87b2a6c809ad15a18e98dd58R11
